core/arch/arm/kernel/ree_fs_ta_buffered.c

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2019, Linaro Limited
+ */
+/*
+ * This is a wrapper around the "REE FS" TA store.
+ * The whole TA/library is read into a temporary buffer during .open(). This
+ * allows the binary to be authenticated before any data is read and processed
+ * by the upper layer (ELF loader).
+ */
+#include <kernel/user_ta.h>
+#include <mm/mobj.h>
+#include <stdlib.h>
+#include <string.h>
+#include <tee_api_types.h>
+
+#include "elf_load.h"
+#include "ree_fs_ta.h"
+
+struct user_ta_store_handle {
+	struct user_ta_store_handle *h; /* Note: a REE FS TA store handle */
+	size_t ta_size;
+	struct mobj *ta_mobj;
+	size_t offs;
+};
+
+static TEE_Result ta_open(const TEE_UUID *uuid,
+			  struct user_ta_store_handle **h)
+{
+	struct user_ta_store_handle *handle = NULL;
+	void *buf = NULL;
+	TEE_Result res = TEE_SUCCESS;
+
+
+	handle = calloc(1, sizeof(*handle));
+	if (!handle)
+		return TEE_ERROR_OUT_OF_MEMORY;
+	res = ree_fs_ta_open(uuid, &handle->h);
+	if (res)
+		goto err2;
+	res = ree_fs_ta_get_size(handle->h, &handle->ta_size);
+	if (res)
+		goto err;
+	handle->ta_mobj = tee_ta_alloc_sec_ddr(handle->ta_size);
```

Comment:
Allocate using `mobj_mm_alloc(mobj_sec_ddr, handle->ta_size, &tee_mm_sec_ddr)` instead. Since a mobj isn't really needed here we could also do something like:
``` C
mm = tee_mm_alloc(&tee_mm_sec_ddr, handle->ta_size);
if (!mm) {
        res = TEE_ERROR_OUT_OF_MEMORY;
        goto err;
}
buf = phys_to_virt(MEM_AREA_TA_RAM, tee_mm_get_smem(mm));
```
This would have the advantage to not depend on `mobj_mm_alloc()` which I'm planning to remove in a PR after https://github.com/OP-TEE/optee_os/pull/2801

Reply:
I'll try that. Thanks.

Reply:
It works as long as the parameters to `phys_to_virt()` are passed in the correct order :wink: 
Updated in [Review] patch.